[guilib] Smart Playlist Rule Dialog Improvements

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

A few improvements of the smart playlist rule dialog (separate commits for each point)

- added a date format validation for rules of type DATE_FIELD and operators after/before (YYYY-MM-DD as required by the DB) - no change to the date entry dialog, which presents the date in the user's local format.
- correctly parse date rules with a date earlier than Jan 1st, 2000 (which was previously chosen for some reason to mean "invalid date"
- default date changed to the current date, rather than Jan 1st, 2000
- apply the validation to a blank value when the rule type is validated (before: a documented behavior of StringUtils::Split() returns an empty vector for a blank string, resulting in a skip of the validation)
- modified the IsTime validation (ex: movie duration) to reject blank values. They're not valid for any operator.
- apply the validation when creating a new rule, when opening the dialog for existing rules, when changing the field of the rule, when changing the operator (before: validated only when when changing the value)

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Reports in forum of failing smart playlist because of incorrect date values, made possible by the lack of validation.
https://forum.kodi.tv/showthread.php?tid=379453

The dialog also allows the creation of rules with blank values even though the field validation requires one.

idea for future change: different hint texts for the different fields, instead of the single hint text set by the skin.
and of course add more validations. For some types, empty value validity depends on the operator (ex. string is [blank] > makes sense. string begins with [blank] > doesn't make sense)

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

creation of new rule, modification of existing rule.
date type fields, also other validated types such rule creation with invalid date isnew rule, change the field between all types 
## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Foolproof the creation of smart playlist using date rules.
Rules missing values for fields that require one cannot be created anymore.

## Screenshots (if appropriate):

invalid date / invalid date format: text in red, OK disabled
![image](https://github.com/user-attachments/assets/921355ad-99e7-4103-8e4c-b327456a36f2)

relative date: no validation at this time.
![image](https://github.com/user-attachments/assets/9ad90865-baa0-49be-8b4d-5807ed6eceac)

blank value for a validated field that doesn't allow empty values: text in red, OK disabled
![image](https://github.com/user-attachments/assets/e41017f1-0bb0-4dcf-9e57-6790fdfea5fe)

empty duration is no longer valid for any operator
![image](https://github.com/user-attachments/assets/bd79c981-5dbc-4f42-878e-49d693541c5e)

unvalidated field: no change, the value field doesn't turn red, OK is always enabled
![image](https://github.com/user-attachments/assets/09518d86-a1e5-4e61-9674-d76d49a99e33)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[guilib] Add validation for smart playlist rules of date type

CrystalP committed:
[guilib] Support dates earlier than Jan 1st 2000 in smart playlist rules, default to current date.

CrystalP committed:
[guilib] Apply validation to empty smart playlist rule value field

CrystalP committed:
[guilib] Validate smart playlist rule value on dialog open and field change.

ksooo commented on xbmc/playlists/SmartPlayList.cpp:
```diff
@@ -290,6 +291,50 @@ bool CSmartPlaylistRule::ValidateMyRating(const std::string &input, void *data)
   return StringValidation::IsPositiveInteger(input, data) && rating <= 10;
 }
 
+static bool StrToInt(const std::string& str, int& num)
```
You decided against an anonymous namespace by intention?

CrystalP replied:
no, I wasn't aware at the time.

ksooo reviewed (commented):
minors only

CrystalP commented on xbmc/playlists/SmartPlayList.cpp:
```diff
@@ -290,6 +291,50 @@ bool CSmartPlaylistRule::ValidateMyRating(const std::string &input, void *data)
+  if (input[4] != '-' || input[7] != '-')
+    return false;
+
+  int year, month, day;
```
Reworked to use an optional

CrystalP commented on xbmc/dialogs/GUIDialogSmartPlaylistRule.h:
```diff
@@ -36,4 +37,5 @@ class CGUIDialogSmartPlaylistRule :
   KODI::PLAYLIST::CSmartPlaylistRule m_rule;
   bool m_cancelled;
   std::string m_type;
+  CGUIEditControl* m_editControl;
```
I'm conflicted about these things. I put the initialization in the constructor to follow the existing pattern. I think a mix is worse than all in one place.
I could move the 3 of them here but that's noise in the history of the, as I'm not changing anything significant.

ksooo replied:
Coding guidelines are clear on this: https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#114-default-member-initialization

But I leave it up to you.

CrystalP replied:
So they are. Changed.

CrystalP commented:
Applied/replied to suggestions, fixed formatting issue.

CrystalP committed:
[guilib] Consider blank duration an invalid input

CrystalP commented:
No change to previous commits with previous comments integrated

Pondered the copilot comment, which wasn't terribly useful but examining the classes closer let me find a much better solution.
All the ground work was laid by PR#3088 with a validation framework, it only missed a prehistoric-period TODO to validate on input type change.

Added a trivial validation to consider empty durations as invalid (ex. movie duration is [blank] cannot work even if the dialog accepted it)

ksooo commented on xbmc/playlists/SmartPlayList.cpp:
```diff
@@ -292,6 +295,52 @@ bool CSmartPlaylistRule::ValidateMyRating(const std::string &input, void *data)
+{
+std::optional<int> StrToInt(const std::string& str)
+{
+  if (std::any_of(str.begin(), str.end(), [](char c) { return !isdigit(c); }))
```
`std::isdigit`?

ksooo replied:
OK.

ksooo reviewed (approved):
LGTM. only one small nitpick.

neo1973 commented on xbmc/playlists/SmartPlayList.cpp:
```diff
@@ -292,6 +296,63 @@ bool CSmartPlaylistRule::ValidateMyRating(const std::string &input, void *data)
+namespace
+{
+template<class T>
+std::optional<T> DigitsToNumeric(const std::string_view& str)
```
The function body is fine as is, but please [pass `std::string_view` by value](https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/).

As for the function: [`std::from_chars`](https://en.cppreference.com/w/cpp/utility/from_chars) could do all the validations for you but it's also fine as is.

CrystalP replied:
yes thanks, I just found out about from_chars and was rewritting.

CrystalP replied:
pushed an update, that's what you had in mind?

neo1973 replied:
Yes, exactly :slightly_smiling_face:

ksooo commented on xbmc/playlists/SmartPlayList.cpp:
```diff
@@ -27,8 +28,12 @@
 #include "utils/XMLUtils.h"
 #include "utils/log.h"
 
+//#include <algorithm>
```
Please don't forget to remove this line.

ksooo commented on xbmc/playlists/SmartPlayList.cpp:
```diff
@@ -27,8 +28,12 @@
+//#include <algorithm>
+#include <charconv>
 #include <cstdlib>
+//#include <limits>
```
Please don't forget to remove this line.

CrystalP commented:
Thanks for the feedback!
Removed the commented lines and squashed the more efficient date validation into the first commit.
Will wait for the green to merge.

neo1973 commented on xbmc/playlists/SmartPlayList.cpp:
```diff
@@ -262,14 +265,14 @@ bool CSmartPlaylistRule::Validate(const std::string &input, void *data)
-    if (!validator(*it, data))
-      return false;
-  }
+  std::vector<std::string> values;
+  if (input.empty())
+    values = {""};
+  else
+    values = StringUtils::Split(input, RULE_VALUE_SEPARATOR);
```
I don't think we need to handle empty input special. `StringUtils::Split` won't insert anything into values and `std::all_of` on an empty range is `true`.

CrystalP replied:
It's an empty element on purpose to run the validation on empty input, the objective of that commit.
I can add a comment to avoid someone removing later by mistake.

CrystalP replied:
Actually I'll call validator explicitly with "" for empty input and there will be no confusion possible ever and no comment needed.

CrystalP commented:
With additional comment clarifying the reason for empty vector element going into the validator.
Will wait for Jenkins to merge.
